### DISCUSSION TITLE
Add a "status" key to useQuery response

### DISCUSSION THREAD

**Discussion by jgoux** [urql-graphql/urql#1456] [Ideas]
Hello 👋 ,

I'd like to have a `status` key returned from useQuery/useMutation so I can easily pattern-match on it to handle every cases (fetching/error/data) exhaustively and explicitly.

Ideally the response type should be a union like : 
```ts
type UseQueryState<Data, Variables> = { status: "fetching" } |  { status: "success", data?: Data } | { status: "error", error?: CombinedError }
```

This is how [react-query](https://react-query.tanstack.com/guides/queries#query-basics) does it. All the other boolean values are just derived from the status as extra helpers.

**Comment by JoviDeCroock** [comment]
This doesn't really work too well for GraphQL as sub-resolvers, ... can error which doesn't mean that there's no data or that it's fatal for the UI. Furthermore we can have partial/stale dat as well.

**Reply by jgoux** [reply]
I'm not sure I understand, my example was a simplification but you do return a finite number of possible combination?

**Reply by kitten** [reply]
The finite number of possible states is actually each permutation of `data` (none, partial, full), `error` (none, GraphQL errors, network w/o data), `fetching`, `stale`. So all states can coexist and hence a state machine like pattern doesn't make sense, as in, if we build it in we'd just encourage people to ignore some permutations even when they shouldn't 😅

**Reply by jgoux** [reply]
You can have both data and error? I didn't know that. ^^

Well then I'll continue my chained `if` conditions I guess! Thanks both of you for the quick feedback!
